File: paykit-mobile/android-demo/app/src/main/java/com/paykit/demo/services/NoiseKeyCache.kt
// NoiseKeyCache.kt
// X25519 Key Cache for Noise Protocol
//
// This file provides caching for derived X25519 keys used in Noise protocol
// communications. Keys are cached both in-memory and persistently in
// EncryptedSharedPreferences to reduce round trips to Pubky Ring.
//
// Key Architecture:
//   - Ed25519 (pkarr) keys are "cold" and managed by Pubky Ring
//   - X25519 (Noise) keys are "hot" and cached locally in Paykit
//   - This cache stores the hot X25519 keys for quick access

package com.paykit.demo.services

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap

/**
 * Cache for X25519 Noise protocol keys
 *
 * This class provides:
 * - In-memory caching for fast access
 * - Persistent storage in EncryptedSharedPreferences for app restarts
 * - Key rotation support via epoch parameter
 * - Secure clearing of keys when needed
 */
class NoiseKeyCache private constructor(context: Context) {
    
    companion object {
        private const val PREFS_NAME = "noise_key_cache"
        private const val KEY_CACHE_INDEX = "noise.key.cache.index"
        
        private fun cacheKey(deviceId: String, epoch: UInt): String {
            return "noise.key.cache.$deviceId.$epoch"
        }
        
        @Volatile
        private var instance: NoiseKeyCache? = null
        
        fun getInstance(context: Context): NoiseKeyCache {
            return instance ?: synchronized(this) {
                instance ?: NoiseKeyCache(context.applicationContext).also { instance = it }
            }
        }
    }
    
    private val prefs: SharedPreferences
    private val memoryCache = ConcurrentHashMap<String, X25519KeypairResult>()
    
    /**
     * Maximum number of keys to keep in cache (per device)
     */
    var maxCachedEpochs: Int = 5
    
    init {
        // Create master key for encryption
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        
        // Create encrypted shared preferences
        prefs = EncryptedSharedPreferences.create(
            context,
            PREFS_NAME,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
        
        loadCacheIndex()
    }
    
    /**
     * Get a cached key if available
     *
     * @param deviceId Device identifier used for derivation
     * @param epoch Key rotation epoch
     * @return Cached keypair or null if not found
     */
    fun getKey(deviceId: String, epoch: UInt): X25519KeypairResult? {
        val key = cacheKey(deviceId, epoch)
        
        // Check memory cache first (fast path)
        memoryCache[key]?.let { return it }
        
        // Check persistent cache
        loadFromPrefs(key)?.let { keypair ->
            // Populate memory cache
            memoryCache[key] = keypair
            return keypair
        }
        
        return null
    }
    
    /**
     * Store a key in the cache
     *
     * @param keypair The keypair to cache
     * @param deviceId Device identifier used for derivation
     * @param epoch Key rotation epoch
     */
    fun setKey(keypair: X25519KeypairResult, deviceId: String, epoch: UInt) {
        val key = cacheKey(deviceId, epoch)
        
        // Store in memory cache
        memoryCache[key] = keypair
        
        // Store in encrypted prefs
        saveToPrefs(keypair, key)
        
        // Update cache index
        updateCacheIndex(add = key)
        
        // Cleanup old epochs if needed
        cleanupOldEpochs(deviceId, epoch)
    }
    
    /**
     * Get or derive a key
     *
     * This method checks the cache first, then derives a new key if needed.
     * Uses PubkyRingIntegration for derivation.
     *
     * @param context Android context
     * @param deviceId Device identifier
     * @param epoch Key rotation epoch
     * @return Keypair (from cache or freshly derived)
     */
    suspend fun getOrDerive(context: Context, deviceId: String, epoch: UInt): X25519KeypairResult {
        // Check cache
        getKey(deviceId, epoch)?.let { return it }
        
        // Derive new key
        val keypair = withContext(Dispatchers.IO) {
            PubkyRingIntegration.getInstance(context).deriveX25519Keypair(deviceId, epoch)
        }
        
        // Cache it
        setKey(keypair, deviceId, epoch)
        
        return keypair
    }
    
    /**
     * Get the latest cached epoch for a device
     *
     * @param deviceId Device identifier
     * @return Latest epoch number or null if no keys cached
     */
    fun getLatestEpoch(deviceId: String): UInt? {
        val prefix = "noise.key.cache.$deviceId."
        var latestEpoch: UInt? = null
        
        for (key in memoryCache.keys) {
            if (key.startsWith(prefix)) {
                val epochStr = key.substringAfterLast('.')
                val epoch = epochStr.toUIntOrNull()
                if (epoch != null && (latestEpoch == null || epoch > latestEpoch)) {
                    latestEpoch = epoch
                }
            }
        }
        
        return latestEpoch
    }
    
    /**
     * Clear a specific key from cache
     *
     * @param deviceId Device identifier
     * @param epoch Key rotation epoch
     */
    fun clearKey(deviceId: String, epoch: UInt) {
        val key = cacheKey(deviceId, epoch)
        
        // Remove from memory
        memoryCache.remove(key)
        
        // Remove from prefs
        prefs.edit().remove(key).apply()
        
        // Update index
        updateCacheIndex(remove = key)
    }
    
    /**
     * Clear all keys for a device
     *
     * @param deviceId Device identifier
     */
    fun clearAllKeys(deviceId: String) {
        val prefix = "noise.key.cache.$deviceId."
        val keysToRemove = memoryCache.keys.filter { it.startsWith(prefix) }
        
        for (key in keysToRemove) {
            memoryCache.remove(key)
            prefs.edit().remove(key).apply()
            updateCacheIndex(remove = key)
        }
    }
    
    /**
     * Clear all cached keys
     */
    fun clearAllKeys() {
        memoryCache.clear()
        prefs.edit().clear().apply()
    }
    
    /**
     * Get cache statistics
     *
     * @return Map with cache statistics
     */
    fun getCacheStats(): Map<String, Any> {
        return mapOf(
            "memoryCount" to memoryCache.size,
            "keys" to memoryCache.keys.toList()
        )
    }
    
    private fun loadCacheIndex() {
        val indexJson = prefs.getString(KEY_CACHE_INDEX, null) ?: return
        
        try {
            val index = JSONArray(indexJson)
            for (i in 0 until index.length()) {
                val key = index.getString(i)
                loadFromPrefs(key)?.let { keypair ->
                    memoryCache[key] = keypair
                }
            }
        } catch (e: Exception) {
            // Ignore corrupted index
        }
    }
    
    private fun loadFromPrefs(key: String): X25519KeypairResult? {
        val json = prefs.getString(key, null) ?: return null
        
        return try {
            val obj = JSONObject(json)
            X25519KeypairResult(
                secretKeyHex = obj.getString("secret_key_hex"),
                publicKeyHex = obj.getString("public_key_hex"),
                deviceId = obj.getString("device_id"),
                epoch = obj.getInt("epoch").toUInt()
            )
        } catch (e: Exception) {
            null
        }
    }
    
    private fun saveToPrefs(keypair: X25519KeypairResult, key: String) {
        val json = JSONObject().apply {
            put("secret_key_hex", keypair.secretKeyHex)
            put("public_key_hex", keypair.publicKeyHex)
            put("device_id", keypair.deviceId)
            put("epoch", keypair.epoch.toInt())
        }
        
        prefs.edit().putString(key, json.toString()).apply()
    }
    
    private fun updateCacheIndex(add: String? = null, remove: String? = null) {
        val index = getCacheIndex().toMutableList()
        
        add?.let {
            if (!index.contains(it)) {
                index.add(it)
            }
        }
        
        remove?.let {
            index.remove(it)
        }
        
        saveCacheIndex(index)
    }
    
    private fun getCacheIndex(): List<String> {
        val indexJson = prefs.getString(KEY_CACHE_INDEX, null) ?: return emptyList()
        
        return try {
            val index = JSONArray(indexJson)
            (0 until index.length()).map { index.getString(it) }
        } catch (e: Exception) {
            emptyList()
        }
    }
    
    private fun saveCacheIndex(index: List<String>) {
        val jsonArray = JSONArray(index)
        prefs.edit().putString(KEY_CACHE_INDEX, jsonArray.toString()).apply()
    }
    
    private fun cleanupOldEpochs(deviceId: String, currentEpoch: UInt) {
        val prefix = "noise.key.cache.$deviceId."
        
        // Get all epochs for this device
        val epochs = mutableListOf<Pair<String, UInt>>()
        
        for (key in memoryCache.keys) {
            if (key.startsWith(prefix)) {
                val epochStr = key.substringAfterLast('.')
                val epoch = epochStr.toUIntOrNull()
                if (epoch != null) {
                    epochs.add(key to epoch)
                }
            }
        }
        
        // Sort by epoch descending
        epochs.sortByDescending { it.second }
        
        // Remove old epochs beyond limit
        while (epochs.size > maxCachedEpochs) {
            val oldest = epochs.removeLast()
            memoryCache.remove(oldest.first)
            prefs.edit().remove(oldest.first).apply()
            updateCacheIndex(remove = oldest.first)
        }
    }
}

File: paykit-mobile/ios-demo/PaykitDemo/PaykitDemo/PaykitDemo/Services/NoiseKeyCache.swift
// NoiseKeyCache.swift
// X25519 Key Cache for Noise Protocol
//
// This file provides caching for derived X25519 keys used in Noise protocol
// communications. Keys are cached both in-memory and persistently in the
// Keychain to reduce round trips to Pubky Ring.
//
// Key Architecture:
//   - Ed25519 (pkarr) keys are "cold" and managed by Pubky Ring
//   - X25519 (Noise) keys are "hot" and cached locally in Paykit
//   - This cache stores the hot X25519 keys for quick access

import Foundation

/// Cache for X25519 Noise protocol keys
///
/// This class provides:
/// - In-memory caching for fast access
/// - Persistent storage in Keychain for app restarts
/// - Key rotation support via epoch parameter
/// - Secure clearing of keys when needed
public final class NoiseKeyCache {
    
    // MARK: - Singleton
    
    public static let shared = NoiseKeyCache()
    
    // MARK: - Constants
    
    private enum Keys {
        static func cacheKey(deviceId: String, epoch: UInt32) -> String {
            return "noise.key.cache.\(deviceId).\(epoch)"
        }
        static let cacheIndex = "noise.key.cache.index"
    }
    
    // MARK: - Properties
    
    private let keychain: KeychainStorage
    private var memoryCache: [String: X25519KeypairResult] = [:]
    private let cacheQueue = DispatchQueue(label: "com.paykit.noise.cache", attributes: .concurrent)
    
    /// Maximum number of keys to keep in cache (per device)
    public var maxCachedEpochs: Int = 5
    
    // MARK: - Initialization
    
    private init() {
        self.keychain = KeychainStorage(serviceIdentifier: Bundle.main.bundleIdentifier ?? "com.paykit.demo")
        loadCacheIndex()
    }
    
    // MARK: - Public Methods
    
    /// Get a cached key if available
    ///
    /// - Parameters:
    ///   - deviceId: Device identifier used for derivation
    ///   - epoch: Key rotation epoch
    /// - Returns: Cached keypair or nil if not found
    public func getKey(deviceId: String, epoch: UInt32) -> X25519KeypairResult? {
        let key = Keys.cacheKey(deviceId: deviceId, epoch: epoch)
        
        // Check memory cache first (fast path)
        var result: X25519KeypairResult?
        cacheQueue.sync {
            result = memoryCache[key]
        }
        
        if let cached = result {
            return cached
        }
        
        // Check persistent cache
        if let keypair = loadFromKeychain(key: key) {
            // Populate memory cache
            cacheQueue.async(flags: .barrier) {
                self.memoryCache[key] = keypair
            }
            return keypair
        }
        
        return nil
    }
    
    /// Store a key in the cache
    ///
    /// - Parameters:
    ///   - keypair: The keypair to cache
    ///   - deviceId: Device identifier used for derivation
    ///   - epoch: Key rotation epoch
    public func setKey(_ keypair: X25519KeypairResult, deviceId: String, epoch: UInt32) {
        let key = Keys.cacheKey(deviceId: deviceId, epoch: epoch)
        
        // Store in memory cache
        cacheQueue.async(flags: .barrier) {
            self.memoryCache[key] = keypair
        }
        
        // Store in keychain
        saveToKeychain(keypair: keypair, key: key)
        
        // Update cache index
        updateCacheIndex(add: key)
        
        // Cleanup old epochs if needed
        cleanupOldEpochs(deviceId: deviceId, currentEpoch: epoch)
    }
    
    /// Get or derive a key
    ///
    /// This method checks the cache first, then derives a new key if needed.
    /// Uses PubkyRingIntegration for derivation.
    ///
    /// - Parameters:
    ///   - deviceId: Device identifier
    ///   - epoch: Key rotation epoch
    /// - Returns: Keypair (from cache or freshly derived)
    public func getOrDerive(deviceId: String, epoch: UInt32) async throws -> X25519KeypairResult {
        // Check cache
        if let cached = getKey(deviceId: deviceId, epoch: epoch) {
            return cached
        }
        
        // Derive new key
        let keypair = try await PubkyRingIntegration.shared.deriveX25519Keypair(
            deviceId: deviceId,
            epoch: epoch
        )
        
        // Cache it
        setKey(keypair, deviceId: deviceId, epoch: epoch)
        
        return keypair
    }
    
    /// Get the latest cached epoch for a device
    ///
    /// - Parameter deviceId: Device identifier
    /// - Returns: Latest epoch number or nil if no keys cached
    public func getLatestEpoch(deviceId: String) -> UInt32? {
        var latestEpoch: UInt32?
        
        cacheQueue.sync {
            let prefix = "noise.key.cache.\(deviceId)."
            for key in memoryCache.keys {
                if key.hasPrefix(prefix),
                   let epochStr = key.components(separatedBy: ".").last,
                   let epoch = UInt32(epochStr) {
                    if latestEpoch == nil || epoch > latestEpoch! {
                        latestEpoch = epoch
                    }
                }
            }
        }
        
        return latestEpoch
    }
    
    /// Clear a specific key from cache
    ///
    /// - Parameters:
    ///   - deviceId: Device identifier
    ///   - epoch: Key rotation epoch
    public func clearKey(deviceId: String, epoch: UInt32) {
        let key = Keys.cacheKey(deviceId: deviceId, epoch: epoch)
        
        // Remove from memory
        cacheQueue.async(flags: .barrier) {
            self.memoryCache.removeValue(forKey: key)
        }
        
        // Remove from keychain
        try? keychain.delete(key: key)
        
        // Update index
        updateCacheIndex(remove: key)
    }
    
    /// Clear all keys for a device
    ///
    /// - Parameter deviceId: Device identifier
    public func clearAllKeys(for deviceId: String) {
        let prefix = "noise.key.cache.\(deviceId)."
        
        cacheQueue.async(flags: .barrier) {
            let keysToRemove = self.memoryCache.keys.filter { $0.hasPrefix(prefix) }
            for key in keysToRemove {
                self.memoryCache.removeValue(forKey: key)
                try? self.keychain.delete(key: key)
                self.updateCacheIndex(remove: key)
            }
        }
    }
    
    /// Clear all cached keys
    public func clearAllKeys() {
        cacheQueue.async(flags: .barrier) {
            for key in self.memoryCache.keys {
                try? self.keychain.delete(key: key)
            }
            self.memoryCache.removeAll()
        }
        
        try? keychain.delete(key: Keys.cacheIndex)
    }
    
    /// Get cache statistics
    ///
    /// - Returns: Dictionary with cache statistics
    public func getCacheStats() -> [String: Any] {
        var stats: [String: Any] = [:]
        
        cacheQueue.sync {
            stats["memoryCount"] = memoryCache.count
            stats["keys"] = Array(memoryCache.keys)
        }
        
        return stats
    }
    
    // MARK: - Private Methods
    
    private func loadCacheIndex() {
        guard let indexData = try? keychain.retrieve(key: Keys.cacheIndex),
              let index = try? JSONDecoder().decode([String].self, from: indexData) else {
            return
        }
        
        // Load all indexed keys into memory cache
        for key in index {
            if let keypair = loadFromKeychain(key: key) {
                memoryCache[key] = keypair
            }
        }
    }
    
    private func loadFromKeychain(key: String) -> X25519KeypairResult? {
        guard let data = try? keychain.retrieve(key: key) else {
            return nil
        }
        
        return try? JSONDecoder().decode(X25519KeypairResult.self, from: data)
    }
    
    private func saveToKeychain(keypair: X25519KeypairResult, key: String) {
        guard let data = try? JSONEncoder().encode(keypair) else {
            return
        }
        
        try? keychain.store(key: key, data: data)
    }
    
    private func updateCacheIndex(add key: String) {
        var index = getCacheIndex()
        if !index.contains(key) {
            index.append(key)
            saveCacheIndex(index)
        }
    }
    
    private func updateCacheIndex(remove key: String) {
        var index = getCacheIndex()
        index.removeAll { $0 == key }
        saveCacheIndex(index)
    }
    
    private func getCacheIndex() -> [String] {
        guard let data = try? keychain.retrieve(key: Keys.cacheIndex),
              let index = try? JSONDecoder().decode([String].self, from: data) else {
            return []
        }
        return index
    }
    
    private func saveCacheIndex(_ index: [String]) {
        guard let data = try? JSONEncoder().encode(index) else {
            return
        }
        try? keychain.store(key: Keys.cacheIndex, data: data)
    }
    
    private func cleanupOldEpochs(deviceId: String, currentEpoch: UInt32) {
        let prefix = "noise.key.cache.\(deviceId)."
        
        cacheQueue.async(flags: .barrier) {
            // Get all epochs for this device
            var epochs: [(key: String, epoch: UInt32)] = []
            
            for key in self.memoryCache.keys {
                if key.hasPrefix(prefix),
                   let epochStr = key.components(separatedBy: ".").last,
                   let epoch = UInt32(epochStr) {
                    epochs.append((key, epoch))
                }
            }
            
            // Sort by epoch descending
            epochs.sort { $0.epoch > $1.epoch }
            
            // Remove old epochs beyond limit
            while epochs.count > self.maxCachedEpochs {
                let oldest = epochs.removeLast()
                self.memoryCache.removeValue(forKey: oldest.key)
                try? self.keychain.delete(key: oldest.key)
                self.updateCacheIndex(remove: oldest.key)
            }
        }
    }
}

